crates/walrus-service/src/client/daemon/routes.rs

```diff
@@ -351,6 +425,225 @@ impl From<ClientError> for GetBlobError {
+                        ClientErrorKind::BlobIdBlocked(_) => ConcatBlobError::BlobBlocked {
+                            blob_id: blob_id.to_string(),
+                        },
+                        ClientErrorKind::BlobTooLarge(size) => ConcatBlobError::BlobTooLarge {
+                            blob_id: blob_id.to_string(),
+                            size: *size,
+                        },
+                        _ => ConcatBlobError::BlobReadFailed {
+                            blob_id: blob_id.to_string(),
+                            message: error.to_string(),
+                        },
+                    };
+                    Err(concat_error)
+                }
+            }
+        }
+    });
+
+    // Convert stream to body
+    let body = Body::from_stream(stream);
+
+    // Build response with appropriate headers
+    let mut response = Response::builder()
+        .status(StatusCode::OK)
+        .header(CONTENT_TYPE, "application/octet-stream")
+        .header(CACHE_CONTROL, "max-age=86400, stale-while-revalidate=3600")
+        .header(X_CONTENT_TYPE_OPTIONS, "nosniff")
+        .body(body)
+        .expect("failed to build concatenated blob response");
+
+    // Populate response headers with priority:
+    // 1. Request header override (if Content-Type in request)
+    // 2. First blob's attributes (if available)
+    // 3. Default (already set above)
+    if let Some(content_type) = request_headers.get(CONTENT_TYPE) {
+        response
+            .headers_mut()
+            .insert(CONTENT_TYPE, content_type.clone());
+    } else if let Some(attribute) = first_attribute {
+        populate_response_headers_from_attributes(
+            response.headers_mut(),
+            &attribute,
+            Some(&response_header_config.allowed_headers),
+        );
+    }
+
+    response
+}
+
+/// Concatenate and stream multiple blobs via GET.
+///
+/// Retrieves multiple blobs specified by comma-separated blob IDs or object IDs in the query
+/// parameter, and streams them concatenated in the order specified. Each blob is loaded,
+/// streamed, and then freed from memory before the next blob is processed.
+#[tracing::instrument(level = Level::ERROR, skip_all)]
+#[utoipa::path(
+    get,
+    path = BLOB_CONCAT_ENDPOINT,
+    params(ConcatQueryParams),
+    responses(
+        (status = 200, description = "Blobs concatenated and streamed successfully", body = [u8]),
+        ConcatBlobError,
+        InvalidConsistencyCheck,
+    ),
+)]
+pub(super) async fn get_blobs_concat<T: WalrusReadClient + Send + Sync + 'static>(
+    State((client, response_header_config)): State<(Arc<T>, Arc<AggregatorResponseHeaderConfig>)>,
+    request_headers: HeaderMap,
+    Query(params): Query<ConcatQueryParams>,
+) -> Response {
+    // Parse consistency check options
+    let consistency_check = match params.consistency_check() {
+        Ok(check) => check,
+        Err(error) => return error.into_response(),
+    };
```

Comment:
Suggestion: Put this into the `concat_blobs_impl`.

Reply:
Done.